stl/src/atomic.cpp

```diff
@@ -33,4 +39,281 @@ _CRTIMP2_PURE void __cdecl _Unlock_shared_ptr_spin_lock() { // release previousl
 #endif // _M_ARM
 }
 
+static std::once_flag _Atomic_spin_count_initialized;
+
+static long _Atomic_spin_count;
+
+static void _Atomic_spin_count_initialize() {
+    _Atomic_spin_count = (std::thread::hardware_concurrency() == 1 ? 0 : 10'000);
+}
+
+enum _Atomic_spin_phase {
+    _ATOMIC_SPIN_PHASE_MASK              = 0xF000'0000,
+    _ATOMIC_WAIT_PHASE_MASK              = 0x0F00'0000,
+    _ATOMIC_SPIN_VALUE_MASK              = 0x00FF'FFFF,
+    _ATOMIC_WAIT_PHASE_SPIN              = 0x0000'0000,
+    _ATOMIC_WAIT_PHASE_WAIT_SET          = 0x0100'0000,
+    _ATOMIC_WAIT_PHASE_WAIT_CLEAR        = 0x0200'0000,
+    _ATOMIC_WAIT_PHASE_WAIT_NO_SEMAPHORE = 0x0300'0000,
+    _ATOMIC_SPIN_PHASE_INIT_SPIN_COUNT   = 0x0000'0000,
+    _ATOMIC_SPIN_PHASE_INIT_SPIN         = 0x1000'0000,
+    _ATOMIC_SPIN_PHASE_INIT_SWITCH_THD   = 0x2000'0000,
+    _ATOMIC_SPIN_PHASE_INIT_SLEEP_ZERO   = 0x3000'0000,
+    _ATOMIC_SPIN_PHASE_INIT_SLEEP        = 0x4000'0000,
+
+    _ATOMIC_SPIN_MASK = _ATOMIC_SPIN_PHASE_MASK | _ATOMIC_SPIN_VALUE_MASK,
+};
+
+bool __cdecl _Atomic_spin_active_only(long& _Spin_context) {
+    switch (_Spin_context & _ATOMIC_SPIN_PHASE_MASK) {
+    case _ATOMIC_SPIN_PHASE_INIT_SPIN_COUNT:
+        std::call_once(_Atomic_spin_count_initialized, _Atomic_spin_count_initialize);
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SPIN + _Atomic_spin_count;
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SPIN:
+        if ((_Spin_context & _ATOMIC_SPIN_VALUE_MASK) > 0) {
+            _Spin_context -= 1;
+            YieldProcessor();
+            return true;
+        }
+    }
+    return false;
+}
+
+void __cdecl _Atomic_spin(long& _Spin_context) {
+    switch (_Spin_context & _ATOMIC_SPIN_PHASE_MASK) {
+    case _ATOMIC_SPIN_PHASE_INIT_SPIN_COUNT:
+        std::call_once(_Atomic_spin_count_initialized, _Atomic_spin_count_initialize);
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SPIN + _Atomic_spin_count;
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SPIN:
+        if ((_Spin_context & _ATOMIC_SPIN_VALUE_MASK) > 0) {
+            _Spin_context -= 1;
+            YieldProcessor();
+            return;
+        }
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SWITCH_THD | (_Spin_context & _ATOMIC_WAIT_PHASE_MASK);
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SWITCH_THD:
+        if (_Spin_context < (_ATOMIC_SPIN_PHASE_INIT_SWITCH_THD + 4)) {
+            _Spin_context += 1;
+            ::SwitchToThread();
+            return;
+        }
+        _Spin_context = _ATOMIC_SPIN_PHASE_INIT_SLEEP_ZERO | (_Spin_context & _ATOMIC_WAIT_PHASE_MASK);
+        [[fallthrough]];
+
+    case _ATOMIC_SPIN_PHASE_INIT_SLEEP_ZERO:
+        if (_Spin_context < (_ATOMIC_SPIN_PHASE_INIT_SLEEP_ZERO + 16)) {
+            _Spin_context += 1;
+            ::Sleep(0);

```

Comment:
To my knowledge, `SwitchToThread` may fail if there's no other ready thread to switch to. `Sleep(0)` will not fail, it will ultimately sacrifice the remaining of time slice if no thread to switch to. 

So there's a [recommendation](http://www.1024cores.net/home/lock-free-algorithms/tricks/spinning) to try `SwitchToThread` before `Sleep(0)`, and I meant to follow it here. If you think it is wrong, I'm fine with that, which one to keep then?

Reply:
I think that's wrong, I just looked at the source code for Sleep(0) (Yes, this is cheating) and it looks like Sleep does "if 0, NtYieldExecution", whereas SwitchToThread just unconditionally calls NtYieldExecution. Not that I've actually traced through this...

Reply:
Assuming you also confirm that `Sleep(0)` returns right after `NtYieldExecution()`, without further branching depending on the value returned, then the difference between `Sleep(0)` and `SwitchToThread` is either a myth from the beginning or something that is no longer true.

Anyway, please comment on XP Support and SRW Lock use. Maybe I'll remove this entire piece instead of removing just `Sleep(0)`

Reply:
Will use SRW Lock as last resort options, so there's no need in `Sleep(0)` or `SwitchToThread` at all

Reply:
@BillyONeal , my test on Windows 7 shows that `Sleep(0)` calls `NtDelayExecution` whereas `SwitchToThread` calls `SwitchToThread` calls `NtYieldExecution()`. Documentation on [Sleep](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep) says about behavior change starting Windows 2003.  

Sure it still does not prove that there is a significant difference between `Sleep(0)` and `SwitchToThread`, but it proves that looking into a source of those functions is not enough, as behavior has changed over history.

Though it is offtopic already, as timed backoff is not needed for CV + SRW Lock implementation.

Reply:
> whereas `SwitchToThread` calls `NtYieldExecution()`

Which does
![image](https://user-images.githubusercontent.com/1544943/76377153-5f1edb00-6307-11ea-83ad-c5cace480963.png)

Reply:
Thanks, then the resolution should probably be using `SwitchToThread` and then `Sleep(nonzero)` next time I'll do backoff.